cpp/src/spatial/haversine.cu

```diff
@@ -0,0 +1,153 @@
+// /*
+//  * Copyright (c) 2020, NVIDIA CORPORATION.
+//  *
+//  * Licensed under the Apache License, Version 2.0 (the "License");
+//  * you may not use this file except in compliance with the License.
+//  * You may obtain a copy of the License at
+//  *
+//  *     http://www.apache.org/licenses/LICENSE-2.0
+//  *
+//  * Unless required by applicable law or agreed to in writing, software
+//  * distributed under the License is distributed on an "AS IS" BASIS,
+//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  * See the License for the specific language governing permissions and
+//  * limitations under the License.
+//  */
+
+#include <cmath>
+#include <memory>
+#include <type_traits>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cuspatial/error.hpp>
+#include "rmm/mr/device/device_memory_resource.hpp"
+#include "rmm/thrust_rmm_allocator.h"
+#include <cuspatial/constants.hpp>
+
+namespace {
+
+template<typename T>
+__device__ T calculate_haversine_distance(T radius, T a_lon, T a_lat, T b_lon, T b_lat) {
+    auto ax = a_lon * DEGREE_TO_RADIAN;
+    auto ay = a_lat * DEGREE_TO_RADIAN;
+    auto bx = b_lon * DEGREE_TO_RADIAN;
+    auto by = b_lat * DEGREE_TO_RADIAN;
+
+    // haversine formula
+    auto x = (bx - ax) / 2;
+    auto y = (by - ay) / 2;
+    auto sinysqrd = std::sin(y)  * std::sin(y);
+    auto sinxsqrd = std::sin(x)  * std::sin(x);
+    auto scale    = std::cos(ay) * std::cos(by);
+
+    return 2 * radius * std::asin(std::sqrt(sinysqrd + sinxsqrd * scale));
+};
+
+struct haversine_functor
+{
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("haversine_distance does not support non-floating-point types.");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(cudf::column_view const& a_lon,
+               cudf::column_view const& a_lat,
+               cudf::column_view const& b_lon,
+               cudf::column_view const& b_lat,
+               double radius,
+               rmm::mr::device_memory_resource* mr,
+               cudaStream_t stream)
+    {
+        auto size = a_lon.size();
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto result = cudf::make_fixed_width_column(cudf::data_type{ tid },
+                                                    size,
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+
+        if (size == 0)
+        {
+            return result;
+        }
```

Comment:
libcudf copying utils can simplify this a lot.

```suggestion
        if (a_lon.is_empty()) return cudf::experimental::empty_like(a_lon);
        auto result = cudf::experimental::detail::allocate_like(a_lon, mask_allocation_policy::NEVER, mr, stream);
```

Reply:
fwiw, I've been doing it this way because it's less overhead cognitively. When I had just started learning the code base, it seemed like there was more than one way to do something, and none of them seemed particularly suitable over the other for single-depth columns. `make_fixed_width_column` was almost always the common denominator for the different ways to _create a fixed width column_.

If I were designing the column factory APIs from scratch, I'd advocate for moving all of the "create a flat fixed width column" logic in to a single method.
1) call it via a template, or give it a data_type. template forwards to data_type version.
2) if the size is zero, or the host-data iterator distance is zero, create an empty column and return.
3) otherwise, create a buffer with the host data, or if there is no host data, create the column with the requested size (defaulting to unallocated, unless indicated otherwise by parameters).
4) optionally create a null mask - from host data if desired.
5) return.

This would be the pattern everywhere, essentially, reducing the overhead of seeing all the different ways it can be done (the more pattern match hits your brain has, the faster you grok code, assuming not coding errors). I have yet to find a place where this pattern is not the simplest to read and understand _imho_. 
```c++
// multiple versions of this `make_column` api would be helpful, but the rest of the pattern is the same, from what I've seen.
auto result = cudf::make_column(a_lon.type(), a_lon.size(), stream, mr);

// would like to see this function added. currently only on `column_view`, etc.
if (result->is_empty())
{
    return result;
}

if (input_determines_nullable_output)
{
    // add a null mask, populate results
    // populate data. optionally skipping writes for invalid values.
}

return result;
```

And while I do have strong opinions about it, I do realize we have those APIs and I'm happy to use them if that's the consensus among the team.

Reply:
That is to say, I see there are fewer arguments passed in the suggested code, but I would argue those parameters are an artifact of the column factory APIs influence on this functor's implementation, and not symptomatic of the functor itself.

Reply:
Your reply is causing me a lot of cognitive overhead. :) Will you make the change or not?